src/test/java/hudson/plugins/copyartifact/CopierTest.java

```diff
@@ -0,0 +1,120 @@
+ */
+public class CopierTest {
+
+    @Test
+    public void test_init() throws IOException, InterruptedException {
+        Run runSrc = Mockito.mock(Run.class);
+        AbstractBuild abDst = Mockito.mock(AbstractBuild.class);
+
+        // make sure legacy parameter types (AbstractBuild) work on all impl variants
+        for (C c : new C[] {new C1(), new C2(), new C3(), new C4()}) {
+            c.init(runSrc, abDst, null, null);
+            Assert.assertEquals(1, c.callCnt);
+        }
+
+        // test with new param types (Run dst type)
+
+        // C1 and C2 both impl the non-legacy init methods, so should work fine.
+        testRunDst(new C1(), 1); // should work
+        testRunDst(new C2(), 1); // should work
+
+        // C3 and C4 only implement the legacy AbstractBuild init method (and not the Run
+        // version). Copier.init(Run, Run) will get called in both cases, but should
+        // not forward the call to C3/C4.init(Run, AbstractBuild) because the supplied runDst
+        // is not an AbstractBuild instance.
+        testRunDst(new C3(), 0);
+        testRunDst(new C4(), 0);
+    }
+
+    public void testRunDst(C c, int expectedCnt) throws IOException, InterruptedException {
+        Run runSrc = Mockito.mock(Run.class);
+        Run runDst = Mockito.mock(Run.class);
+        c.init(runSrc, runDst, null, null);
+        Assert.assertEquals(expectedCnt, c.callCnt);
+    }
+
+    private class C extends Copier {
+        int callCnt;
+        @Override
+        public void copyOne(FilePath source, FilePath target, boolean fingerprintArtifacts) throws IOException, InterruptedException {
+        }
+        @Override
+        public Copier clone() {
+            return null;
+        }
+    }
+
+    private class C1 extends C {
+        // Overrides both init methods (for whatever reason :) )
+        @Override
+        public void init(Run src, Run<?, ?> dst, FilePath srcDir, FilePath baseTargetDir) throws IOException, InterruptedException {
+            callCnt++;
+        }
+
+        @Override
+        public void init(Run src, AbstractBuild<?, ?> dst, FilePath srcDir, FilePath baseTargetDir) throws IOException, InterruptedException {
+            callCnt++;
+        }
+    }
+
+    private class C2 extends C {
+        // Override init that uses the Run dst
+        @Override
+        public void init(Run src, Run<?, ?> dst, FilePath srcDir, FilePath baseTargetDir) throws IOException, InterruptedException {
+            callCnt++;
+        }
+    }
+
+    private class C3 extends C {
+        // Override init that uses the deprecated AbstractBuild dst
+        @Override
+        public void init(Run src, AbstractBuild<?, ?> dst, FilePath srcDir, FilePath baseTargetDir) throws IOException, InterruptedException {
+            callCnt++;
+        }
+    }
+
+    private class C4 extends C {
+        // Override init that uses the deprecated AbstractBuild dst + call super
```

Comment:
Should also test overriding the new overload and calling `super`. There is no reason to do so, of course (the method is logically abstract) but someone might.

By contrast, the use case you are testing _here_ is pointless for the purposes of ensuring that this change brings no regression: the old method _was_ `abstract`, so no existing implementations could have been calling `super`.